### ISSUE TITLE
Inconsistent Training Results

### ISSUE THREAD

**Issue by ayush055** [ultralytics/yolov5#6222]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi,

I am trying to train a YOLOv5s but it seems that I tend to get significantly different results every time. I am using the following command:

python train.py --img 1280 --rect --batch 8 --epochs 10 --data dataset/dataset.yaml --weights yolov5s.pt --hyp yolov5-hyperparameters/hyp.finetune-yolov5s.yaml

The hyperparameters being used are ones that I found through evolution.

First, I trained the model for 10 epochs and had a mAP@0.5:0.95 of **0.401**. When I retrained the model using the same command, the mAP@0.5:0.95 at the end of 10 epochs was **0.383**.

I am wondering if this is normal and what is causing a difference if training in both cases begins with the same pretrained weights.

**Comment by glenn-jocher** [comment]
@ayush055 👋 Hello! Thanks for asking about **training reproducibility**. YOLOv5 🚀 uses a single training seed which is set here using the `init_seeds()` function.
https://github.com/ultralytics/yolov5/blob/6865d19a92d8c160c7fc3c92256627dadce1cd1e/train.py#L108

This function sets `python`, `numpy` and `torch` seeds and updates `cudnn` settings:
https://github.com/ultralytics/yolov5/blob/6865d19a92d8c160c7fc3c92256627dadce1cd1e/utils/general.py#L122-L130

To set a new seed for example:
```python
    init_seeds(2 + RANK)  # set seed=2
```

Note that even when using the same seed trainings may produce different results, especially when using CUDA backends. See https://pytorch.org/docs/stable/notes/randomness.html for details on factors affecting PyTorch training reproducibility and sources of randomness.

<img width="949" alt="Screen Shot 2022-01-08 at 11 20 59 AM" src="https://user-images.githubusercontent.com/26833433/148660366-9155dfd9-351c-4174-818f-2b1b907b3e4e.png">

Good luck 🍀 and let us know if you have any other questions!

**Comment by hugh-98** [comment]
I have the same problem. Have you found a solution?

**Comment by iceisfun** [comment]
If your data set is large enough with correct labels and enough training time you should be very consistent. Most of the time with wild variance in output the data set is small and/or labels are poor from what I have experienced.

What are you changing in the --hyp yolov5-hyperparameters/hyp.finetune-yolov5s.yaml? Until you have the basics working you probably don't need to mess with that

Some people discard the recommendations so its worth giving it a review

[https://docs.ultralytics.com/guides/model-training-tips/](url)

**Comment by pderrenger** [comment]
Thank you for your input! Ensuring a large, well-labeled dataset is crucial for consistent results. If you're experiencing variance, consider reviewing your dataset quality and training parameters. For hyperparameters, sticking to defaults until you achieve stable results is advisable. For more tips, please refer to the [YOLOv5 training tips](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/).
